tour_model_eval/load_predict.py

```diff
@@ -0,0 +1,92 @@
+    model = loadModel(filename)
+    return model
+
+
+def loadModel(filename):
+    fd = open(filename, "r")
+    all_model = fd.read()
+    all_model = jpickle.loads(all_model)
+    fd.close()
+    return all_model
+
+
+def in_bin(bin_location_features,new_trip_location_feat,radius):
+    start_b_lon = new_trip_location_feat[0]
+    start_b_lat = new_trip_location_feat[1]
+    end_b_lon = new_trip_location_feat[2]
+    end_b_lat = new_trip_location_feat[3]
+    for feat in bin_location_features:
+        start_a_lon = feat[0]
+        start_a_lat = feat[1]
+        end_a_lon = feat[2]
+        end_a_lat = feat[3]
+        start = similarity.within_radius(start_a_lat, start_a_lon, start_b_lat, start_b_lon,radius)
+        end = similarity.within_radius(end_a_lat, end_a_lon, end_b_lat, end_b_lon, radius)
+        if start and end:
+            continue
+        else:
+            return False
+    return True
+
+
+def predict_labels(trip):
+    radius = 100
+    user = trip[0]['user_id']
+    trip_feat = preprocess.extract_features(trip)[0]
+    trip_loc_feat = trip_feat[0:4]
+    try:
+        # load locations of bins(1st round of clustering)
+        bin_locations = loadModelStage('locations_' + str(user))[0]
+        # load models from the 2nd round of clustering
+        models = loadModelStage('models_' + str(user))[0]
+        # load user labels in all clusters
+        user_labels = loadModelStage('user_labels_' + str(user))[0]
+    except IOError:
+        return {}
+
+    first_round_label_set = list(bin_locations.keys())
+    sel_fl = None
+    for fl in first_round_label_set:
+        # extract location features of selected bin
+        sel_loc_feat = bin_locations[fl]
+        # Check if start/end locations of the new trip and every start/end locations in this bin are within the range of
+        # radius. If so, the new trip falls in this bin. Then predict the second round label of the new trip
+        # using this bin's model
+        if in_bin(sel_loc_feat, trip_loc_feat, radius):
+            sel_fl = fl
+            break
+    if not sel_fl:
+        return {}
+    # choose selected model
+    sel_model = models[sel_fl]
+    # predict 2nd label for the new trip
+    sel_sl = str(sel_model.predict([trip_feat])[0])
+    # values of the selected 1st round label
+    sel_1st_round_val = user_labels[sel_fl][0]
+    second_label_ls = list(sel_1st_round_val.keys())
+    if sel_sl not in second_label_ls:
+        return {}
+    # values of the selected 2nd round label, wrapped in a list
+    sel_2nd_round_val = sel_1st_round_val[sel_sl]
+
+    return sel_2nd_round_val
+
+
+
+if __name__ == '__main__':
+    participant_uuid_obj = list(edb.get_profile_db().find({"install_group": "participant"}, {"user_id": 1, "_id": 0}))
+    all_users = [u["user_id"] for u in participant_uuid_obj]
+    user = all_users[0]
+    radius = 100
+    trips = preprocess.read_data(user)
+    filter_trips = preprocess.filter_data(trips, radius)
```

Comment:
now you see how it helps to have these functions pulled out into a separate module.